features/steps/component.go

```diff
@@ -2,235 +2,102 @@ package steps
 	cfg, err := config.Get()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get config: %w", err)
+		t.Fatalf("failed to get config: %s", err)
 	}
 
-	cfg.ZebedeeURL = zebedeeURL
-	cfg.Mongo.ClusterEndpoint = mongoAddr
-	cfg.Mongo.Database = utils.RandomDatabase()
-
-	g := &mock.Generator{
-		URLHost: "http://mockhost:9999",
+	component := &Component{
+		ErrorFeature:      componenttest.ErrorFeature{TB: t},
+		AuthFeature:       componenttest.NewAuthorizationFeature(),
+		DatasetFeature:    NewDatasetFeature(t, cfg),
+		CantabularFeature: NewCantabularFeature(),
 	}
+	component.MongoFeature = NewMongoFeature(component.ErrorFeature, g, cfg)
+	component.APIFeature = componenttest.NewAPIFeature(component.Router)
 
-	mongoClient, err := GetWorkingMongo(ctx, cfg, g)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create new mongo mongoClient: %w", err)
-	}
-
-	return &Component{
-		errorChan:  make(chan error),
-		wg:         &sync.WaitGroup{},
-		ctx:        ctx,
-		HTTPServer: &http.Server{},
-		cfg:        cfg,
-		DatasetAPI: httpfake.New(httpfake.WithTesting(t)),
-		CantabularClient: &mock.CantabularClient{
-			OptionsHappy: true,
-		},
-		store:            mongoClient,
-		g:                g,
-		waitEventTimeout: WaitEventTimeout,
-	}, nil
-}
-
-// Init initialises the server, the mocks and waits for the dependencies to be ready
-func (c *Component) Init() (http.Handler, error) {
-	c.signals = make(chan os.Signal, 1)
-	signal.Notify(c.signals, os.Interrupt, syscall.SIGTERM)
-
-	log.Info(c.ctx, "config used by component tests", log.Data{"cfg": c.cfg})
+	cfg.ZebedeeURL = component.AuthFeature.FakeAuthService.ResolveURL("")
 
-	c.cfg.DatasetAPIURL = c.DatasetAPI.ResolveURL("")
+	component.setInitialiserMock(g)
+	component.svc = service.New()
+	component.svc.Cfg = cfg
 
-	// Create service and initialise it
-	c.svc = service.New()
-	if err := c.svc.Init(c.ctx, c.cfg, BuildTime, GitCommit, Version); err != nil {
-		return nil, fmt.Errorf("failed to initialise service: %w", err)
-	}
-
-	return c.HTTPServer.Handler, nil
+	return component
 }
 
-func (c *Component) setInitialiserMock() {
-	service.GetHTTPServer = func(bindAddr string, router http.Handler) service.HTTPServer {
-		c.HTTPServer.Addr = bindAddr
-		c.HTTPServer.Handler = router
-		return c.HTTPServer
-	}
-
-	service.GetGenerator = func() service.Generator {
-		return &mock.Generator{
-			URLHost: "http://mockhost:9999",
-		}
-	}
-
-	service.GetCantabularClient = func(_ *config.Config) service.CantabularClient {
-		return c.CantabularClient
-	}
-
-	service.GetMongoDB = func(ctx context.Context, cfg *config.Config, g service.Generator) (service.Datastore, error) {
-		return c.store, nil
+// Router initialises the service, returning the service's (server) router for tests
+// This delayed initialisation is needed to ensure that any changes to the router (or the service in general)
+// as a result of test setup, are picked up
+func (c *Component) Router() (http.Handler, error) {
```

Comment:
Is this function used anywhere?

Reply:
yes this is the function that is passed to the APIFeature (in func NewComponent():
component.APIFeature = componenttest.NewAPIFeature(component.Router)
and the feature uses it to retrieve the router just before it sends a request to one of the router endpoints. This is the function that ensures the component (the service) is initialised with the correct test setup for the test that the APIFeature is about to call (i.e. in the 'When I Get/Put/Post...' step)